### ISSUE TITLE
inverted colors of images in pdfs and messed OCR after conversion

### ISSUE THREAD

**Issue by maras** [kcroker/dpsprep#16]
Hello.

Dpsprep inverts colors of images (mostly covers of books converted) while converting. Black and white images stays intact, I think, but color and gray-scale (I'm not sure about latter, but I think so) ones always are inverted, so black become white and so on. Sometimes all pages of converted pdf can became black with white letters, but such occasions are rare.
This happens even with minimal command, like `poetry run python -m dpsprep -v`.

As example, colors of the first page of this book (cover) from this [link](http://publ.lib.ru/ARCHIVES/I/Institut_Vseobschey_Istorii_AN/Postijenie_ideala._Iz_istorii_mirotvorchestva_i_intelligencii._Sb.(2005).[djv-fax].zip) (djvu file is zipped and site is http) are inverted. All books from [this library](http://publ.lib.ru/ARCHIVES/I/Institut_Vseobschey_Istorii_AN/_IVIAN.html) can be downloaded and used for personal needs freely and legally. They are in cyrilic, but that should not be problem, I think.

Maybe something is wrong with my system, I don't know. It is latest Manjaro linux (Arch derivative) stable, no Python2, I am working inside recently cloned this repository (didn't install dpsprep as Python module). I can provide more information if it is needed.

Ant thank you for this converter – it is lifesaver when Zotero is working only with pdfs and I have quite a lot of djvu with OCR already. So no need to OCR converted pdfs. And dpsprep is the only converter that preserves OCR, I tried  several solutions.

Thank you,

Valdemaras

**Comment by v--** [comment]
@maras This was happening because in v2 we always inverted the images - djvulibre and Pillow produced inverted monochrome image by themselves. It turns out that color images are read correctly, so I just added conditional logic to invert colors only for monochrome images. See #17.

**Comment by maras** [comment]
Yes, it is working as expected now. Mostly :). Thank you for such a quick fix.

I wrote "mostly" because there are some occasions when empty pages became black after conversion. But only white pages and only in some books, so this issue isn't critical. I uploaded [a book](https://file.io/PtvS7g7AOQpN) (file deletes after one download) conversion from which creates such artifacts. Second, eight, 264 pages are black after conversion. But they are blank so it's not a big problem and I can live with this bug. This file has other issue, but I'll open other ticket for it.

**Comment by v--** [comment]
I will take a look tomorrow, I think I understand what the issue is.

Regarding OCR: You can try running OCRmyPDF on the resulting PDF file and see what comes out.

**Comment by v--** [comment]
@maras Okay, I just merged another fix. The inverted blank pages should be fine now.

PS: I saw that the section titles in "Формы и содержание мышления" were pretty much inadequate, but the main text seemed fine. I may have to debug what goes wrong, but would rather do that as part of a separate issue. Did you notice other problems? Meanwhile, have you tried re-running OCR? The following command should do:
    ocrmypdf --redo-ocr --language rus <file>.pdf <output>.pdf

**Comment by v--** [comment]
I tried better placing for the text boxes in the "hidden" text layer, but I couldn't get it to behave much better for the concrete book, especially for hyphenated words. The problem arises because DjVu assigns text to boxes, while PDF bases it's bounding boxes based on the text in them. We try to resize the (invisible) font and pad our text to fit the same boxes, but there is only so much we can do using this approach.

I may add a note to the documentation recommending OCRmyPDF until somebody comes up with a better implementation (the latter is unlikely to happen). I don't see a point in investing more time into this since you can always re-OCR.

**Comment by maras** [comment]
@v-- Yes, black pages are gone. So this issue is resolved.

Regarding OCR my first intention was to open new ticket but I changed my mind after inspecting original djvu in Ocular, where I saw these layers of boxes one above other (that is page 5 of file): 
![this](https://github.com/kcroker/dpsprep/assets/14958406/30a3091e-707b-45fb-9b07-d00d8cd0f90a)
Text copy form djvu file  works as expected in DJView4 or Ocular, but converted pdf presents such image
![2023-10-28-172259](https://github.com/kcroker/dpsprep/assets/14958406/8ce87cc2-894a-4be7-9ebb-758d33172584)
Copied text from pdf is: 

> предстает Если главным в последующих образом как работах публичный Мамардашвили философ, излагающий
> свои размышления в устной форме (в виде лещищ докла-
> дов, выступлений, интервью), то в этих работах мыписьменной. 
> имеем дело С этим с изначально связан иной ряд особенностей подачей материала, языка этих
> работ, Мамардашвили что сразу по другим отметит публикациям.
>  читатель, знакомый с

 instead of correct from djvu:

> Если в последующих работах Мамардашвили предстает главным образом как публичный философ, излагающий 
> свои размышления в устной форме (в виде лещищ докла- 
> дов, выступлений, интервью), то в этих работах мы
There are permuted words in the first and third lines and copy selected more lines (I copied the same text of first four lines of paragraph in both cases).

After redoing ocr while converting pdf is correct.
Thank you for adding these new options – it eases life!

I'm closing this ticket, because original problem is solved and OCR problem can be solved by re-OCR. And I corrected title of the ticket to be more precise.

Thank you for your work once more :).
